### PR TITLE
fixed an issue where remaining waypoints would always return 0 and break reroute handling

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Fixes a regression from https://github.com/mapbox/mapbox-navigation-android/commit/769430a35084cd102f2b279ee6ce6eee07043113 which caused a crash when a reroute event happens due to a misreported remaining waypoints count, example stacktrace:
```
Process: com.mapbox.navigation.instrumentation_tests, PID: 15480
    java.lang.IllegalStateException: At least 2 coordinates should be provided.
        at com.mapbox.navigation.route.offboard.router.MapboxDirectionsUtilsKt.routeOptions(MapboxDirectionsUtils.kt:20)
        at com.mapbox.navigation.route.internal.offboard.MapboxOffboardRouter.getRoute(MapboxOffboardRouter.kt:53)
        at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter$RouterHandler.handleRouteRequest(MapboxHybridRouter.kt:182)
        at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter$RouterHandler.getRoute(MapboxHybridRouter.kt:160)
        at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter.getRoute(MapboxHybridRouter.kt:205)
        at com.mapbox.navigation.core.directions.session.MapboxDirectionsSession.requestRoutes(MapboxDirectionsSession.kt:81)
        at com.mapbox.navigation.core.reroute.MapboxRerouteController.request(MapboxRerouteController.kt:101)
        at com.mapbox.navigation.core.reroute.MapboxRerouteController.reroute(MapboxRerouteController.kt:58)
        at com.mapbox.navigation.core.MapboxNavigation.reroute(MapboxNavigation.kt:738)
        at com.mapbox.navigation.core.MapboxNavigation.access$reroute(MapboxNavigation.kt:129)
        at com.mapbox.navigation.core.MapboxNavigation$createInternalOffRouteObserver$1.onOffRouteStateChanged(MapboxNavigation.kt:732)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession.setOffRoute(MapboxTripSession.kt:123)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession.access$setOffRoute$p(MapboxTripSession.kt:48)
        at com.mapbox.navigation.core.trip.session.MapboxTripSession$updateDataFromNavigatorStatus$updateNavigatorStatusDataJob$1.invokeSuspend(MapboxTripSession.kt:487)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
        at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
        at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
```

This PR resolves the issue by checking the coordinates count instead of waypoints indices count which reports a different value and adds an instrumentation regression test.

### Changelog

```
<changelog>Fixed a crash when rerouting and fixed a misreported remaining waypoints count that was the root cause.</changelog>
```

### CODE REVIEW

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt [hunk 1]
```diff
@@ -52,6 +53,20 @@ class MockLocationReplayerRule(mockLocationUpdatesRule: MockLocationUpdatesRule)
         mapboxReplayer?.seekTo(replayEvents.first())
         mapboxReplayer?.play()
     }
+
+    fun loopUpdate(location: Location, times: Int) {
+        val events: List<ReplayEventUpdateLocation> =
+            mutableListOf<ReplayEventUpdateLocation>().apply {
+                for (i in 0..times) {
+                    this.add(location.toReplayEventUpdateLocation(i.toDouble()))
+                }
```
Comment (korshaknn) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt:62]:
```suggestion
                repeat(times) {
                    this.add(location.toReplayEventUpdateLocation(i.toDouble()))
                }
```

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt [hunk 2]
```diff
@@ -52,6 +53,20 @@ class MockLocationReplayerRule(mockLocationUpdatesRule: MockLocationUpdatesRule)
...
+            }
+        mapboxReplayer?.stop()
+        mapboxReplayer?.clearEvents()
+        mapboxReplayer?.pushEvents(events)
+        mapboxReplayer?.seekTo(events.first())
+        mapboxReplayer?.play()
```
Comment (korshaknn) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt:68]:
```suggestion
        mapboxReplayer?.run{
            stop()
            clearEvents()
            pushEvents(events)
           seekTo(events.first())
           play()
        }
        
```

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        runOnMainSync {
+            mapboxNavigation.startTripSession()
+            mapboxNavigation.requestRoutes(
+                RouteOptions.builder().applyDefaultParams()
+                    .baseUrl(mockWebServerRule.baseUrl)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .coordinates(mockRoute.routeWaypoints).build()
+            )
+        }
```
Comment (korshaknn) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:109]: wrap with one runOnMainSync{} ?

**Review by korshaknn — APPROVED** [review]
just some nits

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt (hunk 1 above)
Comment (LukasPaczos) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/location/MockLocationReplayer.kt:62]: Ah! I knew Kotlin definitely had something like this :+1:

**Pull request merged**